Sources/tinys3/Helpers.swift

```
@@ -1,24 +1,37 @@
 import Foundation
 import Crypto
 
-struct InvalidDataError: Error {}
-
-struct XMLDataValidator {
-
-    var expectedRootElementName: String
-    var wasTriggered: Bool = false
+enum InvalidXMLResponseError: Error {
+    case invalidRootNode(expected: String, got: String?)
+    case missingElement(String)
+    case unexpectedValue(String, inElement: String)
+}
 
-    mutating func validate(elementName: String, failureHandler: (Error) -> Void) {
-        guard !wasTriggered else {
-            return
+extension XMLDocument {
+    func rootElement(expectedName: String) throws -> XMLElement {
+        let root = self.rootElement()
+        guard let root, root.name == expectedName else {
+            throw InvalidXMLResponseError.invalidRootNode(expected: expectedName, got: root?.name)
         }
+        return root
+    }
+}
 
-        self.wasTriggered = true
-
-        guard elementName == expectedRootElementName else {
-            failureHandler(InvalidDataError())
-            return
+extension XMLElement {
+    func value<T>(forElementName name: String, transform: (String) -> T? = { $0 }) throws -> T {
+        guard let text = self.elements(forName: name).first?.stringValue else {
```

Question: do you think this function should throw an error if there are more than one element for the given name?

---

Good idea. Added in 82c9d038c01756cfd40e59ea2397c8923515f28e